File: app/src/main/java/ru/nikshlykov/englishwordsapp/ui/viewmodels/GroupsViewModel.java
package ru.nikshlykov.englishwordsapp.ui.viewmodels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;

import ru.nikshlykov.englishwordsapp.db.GroupsRepository;
import ru.nikshlykov.englishwordsapp.db.subgroup.Subgroup;
import ru.nikshlykov.englishwordsapp.ui.GroupItem;

public class GroupsViewModel extends AndroidViewModel implements
        GroupsRepository.OnGroupItemsLoadedListener,
        GroupsRepository.OnSubgroupInsertedListener{

    private GroupsRepository groupsRepository;

    private MutableLiveData<ArrayList<GroupItem>> mutableLiveDataGroupItems;

    public GroupsViewModel(@NonNull Application application, GroupsRepository groupsRepository) {
        super(application);
        this.groupsRepository = groupsRepository;

        mutableLiveDataGroupItems = new MutableLiveData<>();
    }

    public MutableLiveData<ArrayList<GroupItem>> getMutableLiveDataGroupItems() {
        return mutableLiveDataGroupItems;
    }

    public void updateSubgroup(Subgroup subgroup) {
        groupsRepository.update(subgroup);
    }

    @Override
    public void onGroupItemsLoaded(ArrayList<GroupItem> groupItems) {
        mutableLiveDataGroupItems.setValue(groupItems);
    }

    @Override
    public void onSubgroupInserted(long subgroupId) {
        // TODO Этот код не будет достигнуть, т.к. теперь Insert делает SubgroupDataFragment.
        //  Поэтому, наверное, после надо будет просто getGroupItems вызывать в ЖЦ GroupsFragment.
        groupsRepository.getGroupItems(this);
    }

    public void loadGroupItems() {
        groupsRepository.getGroupItems(this);
    }
}

File: app/src/main/java/ru/nikshlykov/englishwordsapp/ui/viewmodels/GroupsViewModel.kt
package ru.nikshlykov.englishwordsapp.ui.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import ru.nikshlykov.englishwordsapp.db.models.Subgroup
import ru.nikshlykov.englishwordsapp.domain.interactors.GetGroupsWithSubgroupsInteractor
import ru.nikshlykov.englishwordsapp.domain.interactors.UpdateSubgroupInteractor
import ru.nikshlykov.englishwordsapp.ui.GroupItem

class GroupsViewModel(
  application: Application,
  private val getGroupsWithSubgroupsInteractor: GetGroupsWithSubgroupsInteractor,
  private val updateSubgroupInteractor: UpdateSubgroupInteractor
) :
  AndroidViewModel(application) {
  private val _groupItems: MutableLiveData<ArrayList<GroupItem>?> = MutableLiveData()
  val groupItems: LiveData<ArrayList<GroupItem>?> = _groupItems

  fun updateSubgroup(subgroup: Subgroup?) {
    viewModelScope.launch {
      // TODO убрать потом проверку на null
      if (subgroup != null) {
        updateSubgroupInteractor.updateSubgroup(subgroup)
      }
    }
  }

  fun loadGroupItems() {
    viewModelScope.launch {
      _groupItems.value =
        getGroupsWithSubgroupsInteractor.getAllGroupsWithSubgroups()
    }
  }
}